### PR TITLE
100% Discount cart rule results in Tax invoiced #30853

### LINKED ISSUE #30853: 100% Discount cart rule results in Tax invoiced
### Preconditions (*)

1. Magento 2.3.6
2. Magento 2.4-develop
3. Free shipping is enabled

### Steps to reproduce (*)

1. Navigate to Stores / Settings /  Configuration / Sales / Tax
2. Set Catalog Prices and Apply Discount on Prices to Including Tax 
image
3. Navigate to Sores / Tax Rules / Add New Tax Rule 
4. Create a rule for all USA with 10% tax rate and apply it 
5. Navigate to Marketing / Cart Price Rules and Add New Rule
6. Create a rule with 100% discount for all users 
image
7. Make an order on Storefront:
- choose Free Shipping
- apply coupon code 
8. Navigate to Sales / Orders  and open your order
9. Create an invoice for the order and open it  

### Expected result (*)

1. An invoice is created over 0,00$

### Actual result (*)

1. An invoice with the Grand Total of the tax amount is created
image
Peek 2020-11-12 10-50

I'm suspect this bug was introduced in this commit (MC-35675):
https://github.com/magento/magento2/commit/2dbab41631cce46be4a27dcbf10a7b172be61f63

It doesn't allow GrandTotals to go below zero, but the Tax Totals will add the discounted Tax, expecting that it was deducted before, resulting in a Invoice which is never smaller than the tax amount (instead of being 0)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by engcom-Hotel**
### Description (*)
https://github.com/magento/magento2/issues/30853

### Fixed Issues (if relevant)

1. Fixes https://github.com/magento/magento2/issues/30853

### Manual testing scenarios (*)

1. https://github.com/magento/magento2/issues/30853

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B, Functional Tests EE

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Hotel** [comment]
@magento run Functional Tests B2B

**Comment by gabrieldagama** [comment]
@magento give me 2.4 instance with edition ee

**Comment by barryvdh** [comment]
I've applied this fix but it's still generating Invoices with incorrect values on 2.3.6. Also with 99% discount.

File: app/code/Magento/Sales/Model/Order/Invoice/Total/Tax.php [hunk 1]
```diff
@@ -118,8 +118,13 @@ public function collect(\Magento\Sales\Model\Order\Invoice $invoice)
...
 
-        $invoice->setGrandTotal($invoice->getGrandTotal() + $totalTax + $totalDiscountTaxCompensation);
-        $invoice->setBaseGrandTotal($invoice->getBaseGrandTotal() + $baseTotalTax + $baseTotalDiscountTaxCompensation);
+        $grandTotal = $invoice->getGrandTotal() + $invoice->getDiscountAmount() < 0.0001
+            ? 0 : $invoice->getGrandTotal() + $totalTax + $totalDiscountTaxCompensation;
+        $baseGrandTotal = $invoice->getBaseGrandTotal() + $invoice->getBaseDiscountAmount() < 0.0001
+            ? 0 : $invoice->getBaseGrandTotal() + $baseTotalTax + $baseTotalDiscountTaxCompensation;
+
+        $invoice->setGrandTotal($grandTotal);
+        $invoice->setBaseGrandTotal($baseGrandTotal);
```
Comment (gabrieldagama) [inline app/code/Magento/Sales/Model/Order/Invoice/Total/Tax.php]: Such a solution would just work if the Tax totals is the last to run, but depending on sort order is not great, I recommend looking at `\Magento\Sales\Model\Order\Invoice\Total\Shipping` line 41 and 42, that is where the shipping cost is bein added to the grand total.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]

**Comment by engcom-Foxtrot** [comment]
The solution should be fully reworked. The issue seems to be caused by tax, discount, and discount_tax_compensation_amount incorrect calculation.

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run WebAPI Tests, Functional Tests B2B

**Comment by hostep** [comment]
Before this gets merged, please confirm that this is still needed on the latest `2.4-develop` branch, because I think it doesn't anymore: https://github.com/magento/magento2/issues/30853#issuecomment-831977096

**Comment by engcom-Alfa** [comment]
Hi All, thanks for your contribution over here!

Checked the linked Github issue [#30853](https://github.com/magento/magento2/issues/30853), and it has already been fixed with the help of internal Jira ticket by the internal team.

We have the older commit links of this ticket here - https://github.com/magento/magento2/commit/533ec2eb46c91b79b7a99ba4a6a164e2f6e3d7f1

Hence, closing this PR currently.
In case of any further clarifications on this, please do comment in the same ticket.
If you feel like, require to do more changes/modifications/fixes coverage, you may report a new PR as well.

Thanks!

**Pull request closed without merging**